### ISSUE TITLE
CEF still rebuilding too much stuff on Linux, due to the headless feature

### ISSUE THREAD

**Issue by larsbergstrom** [servo/servo#9922]
On our Linux CI builds, we set `SERVO_HEADLESS=1` and thus pass `--feature=headless` to the main build and unit tests. Unfortunately, when we do `build-cef`, we do NOT pass that feature in. Therefore, as seen in the fingerprint log below, we are forced to rebuild the whole thing. There are a few options here:

1) Hurry up and help #8641 get landed
2) On CI, go ahead and build CEF with the headless feature enabled, as we're mainly just making sure it still builds anyway

We need to do *something* soon, though, as it's burning an extra 16 minutes or so.

Opinions? @metajack @nox 

CC @alexcrichton (though I don't think there's any work to do here for Cargo, just FYI)

```
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for embedding v0.0.1 (file:///home/larsberg/servo/ports/cef): failed to read `/home/larsberg/servo/target/release/.fingerprint/embedding-811bb6732ff26da8/lib-embedding`
INFO:cargo::ops::cargo_rustc::fingerprint:   cause: No such file or directory (os error 2)
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for style_traits v0.0.1 (file:///home/larsberg/servo/ports/cef): new (util v0.0.1 (file:///home/larsberg/servo/ports/cef)) != old (util v0.0.1 (file:///home/larsberg/servo/c\
omponents/servo))
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for net_traits v0.0.1 (file:///home/larsberg/servo/ports/cef): new (msg v0.0.1 (file:///home/larsberg/servo/ports/cef)) != old (msg v0.0.1 (file:///home/larsberg/servo/compo\
nents/servo))
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for msg v0.0.1 (file:///home/larsberg/servo/ports/cef): new (layers v0.2.2 (https://github.com/servo/rust-layers#dd28dbee)) != old (layers v0.2.2 (https://github.com/servo/r\
ust-layers#dd28dbee))
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for layers v0.2.2 (https://github.com/servo/rust-layers#dd28dbee): new (azure v0.4.1 (https://github.com/servo/rust-azure#d318e2d8)) != old (azure v0.4.1 (https://github.com\
/servo/rust-azure#d318e2d8))
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for azure v0.4.1 (https://github.com/servo/rust-azure#d318e2d8): new (servo-skia v0.20130412.6) != old (servo-skia v0.20130412.6)
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for servo-skia v0.20130412.6: new (servo-glutin v0.4.9) != old (servo-glutin v0.4.9)
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for servo-glutin v0.4.9: features have changed: Some(["default"]) != Some(["default", "headless"])
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for servo-glutin v0.4.9: features have changed: Some(["default"]) != Some(["default", "headless"])
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for gfx v0.0.1 (file:///home/larsberg/servo/ports/cef): new (azure v0.4.1 (https://github.com/servo/rust-azure#d318e2d8)) != old (azure v0.4.1 (https://github.com/servo/rust\
-azure#d318e2d8))
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for canvas_traits v0.0.1 (file:///home/larsberg/servo/ports/cef): new (azure v0.4.1 (https://github.com/servo/rust-azure#d318e2d8)) != old (azure v0.4.1 (https://github.com/\
servo/rust-azure#d318e2d8))
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for gfx_traits v0.0.1 (file:///home/larsberg/servo/ports/cef): new (azure v0.4.1 (https://github.com/servo/rust-azure#d318e2d8)) != old (azure v0.4.1 (https://github.com/ser\
vo/rust-azure#d318e2d8))
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for style v0.0.1 (file:///home/larsberg/servo/ports/cef): new (style_traits v0.0.1 (file:///home/larsberg/servo/ports/cef)) != old (style_traits v0.0.1 (file:///home/larsber\
g/servo/components/servo))
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for devtools v0.0.1 (file:///home/larsberg/servo/ports/cef): new (devtools_traits v0.0.1 (file:///home/larsberg/servo/ports/cef)) != old (devtools_traits v0.0.1 (file:///hom\
e/larsberg/servo/components/servo))
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for devtools_traits v0.0.1 (file:///home/larsberg/servo/ports/cef): new (msg v0.0.1 (file:///home/larsberg/servo/ports/cef)) != old (msg v0.0.1 (file:///home/larsberg/servo/\
components/servo))
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for servo v0.0.1 (file:///home/larsberg/servo/ports/cef): features have changed: Some(["default", "glutin_app", "webdriver", "webdriver_server", "window"]) != Some(["glutin_\
app", "headless"])
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for glutin_app v0.0.1 (file:///home/larsberg/servo/ports/cef): features have changed: Some(["window"]) != Some(["headless", "servo-glutin"])
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for compositing v0.0.1 (file:///home/larsberg/servo/ports/cef): new (azure v0.4.1 (https://github.com/servo/rust-azure#d318e2d8)) != old (azure v0.4.1 (https://github.com/se\
rvo/rust-azure#d318e2d8))
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for layout_traits v0.0.1 (file:///home/larsberg/servo/ports/cef): new (gfx v0.0.1 (file:///home/larsberg/servo/ports/cef)) != old (gfx v0.0.1 (file:///home/larsberg/servo/co\
mponents/servo))
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for script_traits v0.0.1 (file:///home/larsberg/servo/ports/cef): new (canvas_traits v0.0.1 (file:///home/larsberg/servo/ports/cef)) != old (canvas_traits v0.0.1 (file:///ho\
me/larsberg/servo/components/servo))
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for canvas v0.0.1 (file:///home/larsberg/servo/ports/cef): new (azure v0.4.1 (https://github.com/servo/rust-azure#d318e2d8)) != old (azure v0.4.1 (https://github.com/servo/r\
ust-azure#d318e2d8))
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for net v0.0.1 (file:///home/larsberg/servo/ports/cef): new (devtools_traits v0.0.1 (file:///home/larsberg/servo/ports/cef)) != old (devtools_traits v0.0.1 (file:///home/lar\
sberg/servo/components/servo))
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for script v0.0.1 (file:///home/larsberg/servo/ports/cef): new (canvas v0.0.1 (file:///home/larsberg/servo/ports/cef)) != old (canvas v0.0.1 (file:///home/larsberg/servo/com\
ponents/servo))
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for script v0.0.1 (file:///home/larsberg/servo/ports/cef): precalculated components have changed: 1457443795.259634685s (/home/larsberg/servo/components/script/parser.out) !\
= 1457400737.206154249s (/home/larsberg/servo/components/script/parser.out)
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for profile v0.0.1 (file:///home/larsberg/servo/ports/cef): new (profile_traits v0.0.1 (file:///home/larsberg/servo/ports/cef)) != old (profile_traits v0.0.1 (file:///home/l\
arsberg/servo/components/servo))
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for webdriver_server v0.0.1 (file:///home/larsberg/servo/ports/cef): failed to read `/home/larsberg/servo/target/release/.fingerprint/webdriver_server-df608261ceb1af05/lib-w\
ebdriver_server`
INFO:cargo::ops::cargo_rustc::fingerprint:   cause: No such file or directory (os error 2)
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for webdriver v0.4.0 (https://github.com/jgraham/webdriver-rust.git#9dffcbe4): failed to read `/home/larsberg/servo/target/release/.fingerprint/webdriver-514d252bfebab441/li\
b-webdriver`
INFO:cargo::ops::cargo_rustc::fingerprint:   cause: No such file or directory (os error 2)
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for layout v0.0.1 (file:///home/larsberg/servo/ports/cef): new (azure v0.4.1 (https://github.com/servo/rust-azure#d318e2d8)) != old (azure v0.4.1 (https://github.com/servo/r\
ust-azure#d318e2d8))
INFO:cargo::ops::cargo_rustc::fingerprint: fingerprint error for servo v0.0.1 (file:///home/larsberg/servo/ports/cef): features have changed: Some(["default", "glutin_app", "webdriver", "webdriver_server", "window"]) != Some(["glutin_\
app", "headless"])
```

**Comment by larsbergstrom** [comment]
Hrm, even if I pass the same features, I still get these errors.

@alexcrichton Should the fingerprint path matching work if there are optional features enabled? I might be doing something wrong here, but even if I try to match them all up, we still get a LOT of rebuilding.

**Comment by larsbergstrom** [comment]
Oh, I might also not be fully threading all features properly through the full tree of our dependencies. I'll give that another try...

**Comment by alexcrichton** [comment]
@larsbergstrom if you set the headless features, what's the output of the `RUST_LOG` annotation? It looks like from the logs above it's not only the `headless` features sometimes being activated or not but it's also the `webdriver` feature for at least the servo crate

**Comment by metajack** [comment]
Must headless be a feature? This used to work with an env var, which was not the best, but it seems reasonable to want to build servo for dev but run tests headless if you are on Linux.

**Comment by larsbergstrom** [comment]
@metajack There is a PR I mentioned above (#8641) that would remove the headless feature.

----------

@alexcrichton I think you're right, looking into it more! I do need to add the other features to the rest of the dependency tree. I'll try that out.

**Comment by larsbergstrom** [comment]
Fixed by #8641
